Add concatenation transformer

brbrown25 opened the pull request:
Closes #20.

brbrown25 commented:
@ivanyu what do you think of this proposed Pr?

brbrown25 commented:
@willyborankin  or @ivanyu would it be possible to get a review on this? Thanks!

ivanyu commented on src/main/java/io/aiven/kafka/connect/transforms/ConcatFieldsConfig.java:
```diff
@@ -0,0 +1,92 @@
+    public static final String FIELD_NAMES_CONFIG = "field.names";
+    private static final String FIELD_NAMES_DOC =
+        "A comma separated list of fields to concatenate.";
+    public static final String FIELD_NAME_CONFIG = "field.name";
```
This config name should be a bit more distinctive than `field.names`. Maybe something like `output.field.name` or `result.field.name`.

brbrown25 replied:
Yeah I really like `output.field.name` I'll go that route. 

- [ ] rename `field.names` to `output.field.name`

willyborankin commented on src/main/java/io/aiven/kafka/connect/transforms/ConcatFields.java:
```diff
@@ -0,0 +1,126 @@
+                        fieldValues.add(struct.get(field));
+                    }
+                } catch (final Exception e) {
+                    log.debug(field + " is missing, concat will use " + config.fieldReplaceMissing());
```
Since we use `SLF4j`, could you please use parameterized messages instead of string concatenation?

brbrown25 replied:
Yep I'll update that accordingly.

- [ ] use parameterized messages instead of string concatenation

willyborankin commented on src/main/java/io/aiven/kafka/connect/transforms/ConcatFields.java:
```diff
@@ -0,0 +1,126 @@
+                fieldValues.stream().map(Object::toString).collect(Collectors.joining(config.delimiter())));
+            newRecord = createNewRecord(record, newSchema.build(), newStruct);
+        } else if (record.value() instanceof Map) {
+            final Map newValue = (Map) record.value();
```
Would be better to convert it to the expected type or `Map<?, ?>` in worst case

brbrown25 replied:
- [ ] convert to `Map<String, ?>`

willyborankin commented on src/main/java/io/aiven/kafka/connect/transforms/ConcatFields.java:
```diff
@@ -0,0 +1,126 @@
+                }
+            });
+            newValue.put(config.fieldName(),
+                fieldValues.stream().map(Object::toString).collect(Collectors.joining(config.delimiter())));
```
Better to use for-loop here instead of streams. `JDK14` contains optimization for such code, but not `JDK11`, which is our default JDK

brbrown25 replied:
Gotcha, I'll switch to the for loop syntax!

- [ ] change from streams to for loop

willyborankin reviewed (changes requested):
Thank you very much for your PR. It looks OK, but besides comments I and @ivan added could you please provide documentation for this feature as well?

brbrown25 commented:
> Thank you very much for your PR. It looks OK, but besides comments I and @ivan added could you please provide documentation for this feature as well?

Yep I'll add in the documentation as well. Should have an update this afternoon

brbrown25 commented:
@willyborankin I've applied all the requested changes except for the `Map<String, ?>` with that I was receiving the following error
```
/Users/brown/git/aiven-kafka-connect-transforms/src/main/java/io/aiven/kafka/connect/transforms/ConcatFields.java:104: error: incompatible types: String cannot be converted to CAP#1
            newValue.put(config.outputFieldName(), outputValue.toString());
                                                                       ^
  where CAP#1 is a fresh type-variable:
    CAP#1 extends Object from capture of ?
Note: /Users/brown/git/aiven-kafka-connect-transforms/src/main/java/io/aiven/kafka/connect/transforms/ConcatFields.java uses unchecked or unsafe operations.
```
I also switched to using StringJoiner at the suggestion of IntelliJ, but can change that if you'd like.

willyborankin commented:
> @willyborankin I've applied all the requested changes except for the `Map<String, ?>` with that I was receiving the following error
> 
> ```
> /Users/brown/git/aiven-kafka-connect-transforms/src/main/java/io/aiven/kafka/connect/transforms/ConcatFields.java:104: error: incompatible types: String cannot be converted to CAP#1
>             newValue.put(config.outputFieldName(), outputValue.toString());
>                                                                        ^
>   where CAP#1 is a fresh type-variable:
>     CAP#1 extends Object from capture of ?
> Note: /Users/brown/git/aiven-kafka-connect-transforms/src/main/java/io/aiven/kafka/connect/transforms/ConcatFields.java uses unchecked or unsafe operations.
> ```
> 
> I also switched to using StringJoiner at the suggestion of IntelliJ, but can change that if you'd like.

Then it should be `Map<?, ?>`

brbrown25 commented:
> > @willyborankin I've applied all the requested changes except for the `Map<String, ?>` with that I was receiving the following error
> > ```
> > /Users/brown/git/aiven-kafka-connect-transforms/src/main/java/io/aiven/kafka/connect/transforms/ConcatFields.java:104: error: incompatible types: String cannot be converted to CAP#1
> >             newValue.put(config.outputFieldName(), outputValue.toString());
> >                                                                        ^
> >   where CAP#1 is a fresh type-variable:
> >     CAP#1 extends Object from capture of ?
> > Note: /Users/brown/git/aiven-kafka-connect-transforms/src/main/java/io/aiven/kafka/connect/transforms/ConcatFields.java uses unchecked or unsafe operations.
> > ```
> > 
> > 
> > I also switched to using StringJoiner at the suggestion of IntelliJ, but can change that if you'd like.
> 
> Then it should be `Map<?, ?>`

Ok I've switched to `Map<?, ?>` for the cast. I still have to keep the return type as `Map`, making it `Map<?, ?>` caused
```
> Task :compileJava FAILED
/Users/brown/git/aiven-kafka-connect-transforms/src/main/java/io/aiven/kafka/connect/transforms/ConcatFields.java:103: error: method put in interface Map<K,V> cannot be applied to given types;
            newValue.put(config.outputFieldName(), outputValue.toString());
                    ^
  required: CAP#1,CAP#2
  found: String,String
  reason: argument mismatch; String cannot be converted to CAP#1
  where K,V are type-variables:
    K extends Object declared in interface Map
    V extends Object declared in interface Map
  where CAP#1,CAP#2 are fresh type-variables:
    CAP#1 extends Object from capture of ?
    CAP#2 extends Object from capture of ?
1 error

FAILURE: Build failed with an exception.
```
Let me know any other changes you'd like and I'll knock them out, thanks!

ivanyu commented on src/main/java/io/aiven/kafka/connect/transforms/ConcatFields.java:
```diff
@@ -0,0 +1,134 @@
+                    } else {
+                        fieldValues.add(struct.get(field));
+                    }
+                } catch (final Exception e) {
```
It'd be nice to not catch so broad an exception class and limit it to `DataException`, which is thrown when a filed is not found.

ivanyu commented on src/main/java/io/aiven/kafka/connect/transforms/ConcatFields.java:
```diff
@@ -0,0 +1,134 @@
+            config.fieldNames().forEach(field -> {
+                try {
+                    if (struct.get(field) == null) {
+                        fieldValues.add(config.fieldReplaceMissing());
```
Is it necessary to separately collect `fieldValues` and then add them to `StringJoiner`? I guess `fieldValues` could be omitted and values go directly to `StringJoiner`.

ivanyu commented on src/main/java/io/aiven/kafka/connect/transforms/ConcatFields.java:
```diff
@@ -0,0 +1,134 @@
+        );
+    }
+
+    public static final class Value<R extends ConnectRecord<R>> extends ConcatFields<R> {
+        // There's an implementation only for value, not for key.
+        // We provide $Value class anyway for the consistency sake
+        // and in case we need a $Key version in the future as well.
+    }
```
The documentation says both classes are implemented. The implementation (+ tests) is needed.

ivanyu commented on src/main/java/io/aiven/kafka/connect/transforms/ConcatFieldsConfig.java:
```diff
@@ -0,0 +1,92 @@
+        return new ConfigDef()
+            .define(
+                FIELD_NAMES_CONFIG,
+                ConfigDef.Type.STRING,
```
Should be `ConfigDef.Type.LIST`.

ivanyu commented on src/main/java/io/aiven/kafka/connect/transforms/ConcatFieldsConfig.java:
```diff
@@ -0,0 +1,92 @@
+            .define(
+                DELIMITER_CONFIG,
+                ConfigDef.Type.STRING,
+                ConfigDef.NO_DEFAULT_VALUE,
```
An empty string `""` might be a good default value here.

ivanyu commented on src/main/java/io/aiven/kafka/connect/transforms/ConcatFieldsConfig.java:
```diff
@@ -0,0 +1,92 @@
+            .define(
+                FIELD_REPLACE_MISSING_CONFIG,
+                ConfigDef.Type.STRING,
+                ConfigDef.NO_DEFAULT_VALUE,
```
An empty string `""` might be a good default value here.

ivanyu commented on src/main/java/io/aiven/kafka/connect/transforms/ConcatFieldsConfig.java:
```diff
@@ -0,0 +1,92 @@
+                DELIMITER_CONFIG,
+                ConfigDef.Type.STRING,
+                ConfigDef.NO_DEFAULT_VALUE,
+                new ConfigDef.NonEmptyString(),
```
An empty string looks like a valid value.

ivanyu commented on src/main/java/io/aiven/kafka/connect/transforms/ConcatFieldsConfig.java:
```diff
@@ -0,0 +1,92 @@
+                FIELD_REPLACE_MISSING_CONFIG,
+                ConfigDef.Type.STRING,
+                ConfigDef.NO_DEFAULT_VALUE,
+                new ConfigDef.NonEmptyString(),
```
An empty string looks like a valid value.

brbrown25 commented:
Awesome I'll work through these comments in the next day or two. Thanks!

brbrown25 commented:
@ivanyu addresses all of your feedback.

ivanyu commented on src/test/java/io/aiven/kafka/connect/transforms/ConcatFieldsTest.java:
```diff
@@ -0,0 +1,271 @@
+        newValueMap.put(FIELD, TEST_VALUE.toString() + "-" + FOO_VALUE + "-" + BAR_VALUE + "-" + AGE_VALUE);
+
+        final SinkRecord originalRecord = record(VALUE_SCHEMA, valueMap);
+        final SinkRecord transformedRecord = transformation().apply(originalRecord);
```
I'm afraid the Key variant effectively doesn't work or works incorrectly. If you run the Key tests, `transformedRecord` will have the value overwritten with what was supposed to be the new key. Below go some pointers.

ivanyu commented on src/test/java/io/aiven/kafka/connect/transforms/ConcatFieldsTest.java:
```diff
@@ -0,0 +1,271 @@
+    private SinkRecord setNewValue(final SinkRecord record, final Schema newValueSchema, final Object newValue) {
+        return record.newRecord(record.topic(),
+            record.kafkaPartition(),
+            record.keySchema(),
+            record.key(),
+            newValueSchema,
+            newValue,
```
The function is used in both tests, but it always sets the value.

ivanyu commented on src/main/java/io/aiven/kafka/connect/transforms/ConcatFields.java:
```diff
@@ -0,0 +1,180 @@
+            newStruct.put(config.outputFieldName(), outputValue.toString());
+            newRecord = createNewRecord(record, newSchema.build(), newStruct);
+        } else if (schemaAndValue.value() instanceof Map) {
+            final Map newValue = (Map<?, ?>) schemaAndValue.value();
```
This mutates the old values in place, which combined with other issues make the tests green, while the code is incorrect. Please make a copy of it like:
```suggestion
            final Map<?, ?> newValue = new HashMap<>((Map<?, ?>) schemaAndValue.value());
```

ivanyu reviewed (changes requested):
Thank you for addressing the comments!
Unfortunately, here are some more. The main issue is that the Key variant of the transformation doesn't work correctly, despite the tests pass (it seems like an interplay of bugs in the test code and the transformation code).

ivanyu commented on src/main/java/io/aiven/kafka/connect/transforms/ConcatFields.java:
```diff
@@ -0,0 +1,180 @@
+        this.config = new ConcatFieldsConfig(configs);
+    }
+
+/*
+        final SchemaAndValue schemaAndValue = getSchemaAndValue(record);
+        if (schemaAndValue.schema() == null) {
+            throw new DataException(dataPlace() + " schema can't be null: " + record);
+        }
+
+        final Optional<Object> newValue;
+        if (config.fieldName().isPresent()) {
+            newValue = getNewValueForNamedField(
+                    record.toString(), schemaAndValue.schema(), schemaAndValue.value(), config.fieldName().get());
+        } else {
+            newValue = getNewValueWithoutFieldName(
+                    record.toString(), schemaAndValue.schema(), schemaAndValue.value());
+        }
+
+        if (newValue.isPresent()) {
+            return record.newRecord(
+                    record.topic(),
+                    record.kafkaPartition(),
+                    record.keySchema(),
+                    record.key(),
+                    record.valueSchema(),
+                    newValue.get(),
+                    record.timestamp(),
+                    record.headers()
+            );
+        } else {
+            return record;
+        }
+ */
```
Maybe something needs to be done to this leftover

brbrown25 replied:
haha oops yeah I meant to delete that!

ivanyu commented on src/test/java/io/aiven/kafka/connect/transforms/ConcatFieldsTest.java:
```diff
@@ -0,0 +1,271 @@
+
+        final SinkRecord originalRecord = record(VALUE_SCHEMA, valueMap);
+        final SinkRecord transformedRecord = transformation().apply(originalRecord);
+        assertEquals(setNewValue(originalRecord, NEW_VALUE_SCHEMA, newValueMap).toString(),
```
No need to call `toString()`, `SinkRecords` are comparable (in some other places too).

brbrown25 commented on src/main/java/io/aiven/kafka/connect/transforms/ConcatFields.java:
```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright 2019 Aiven Oy
```
Of course!

brbrown25 commented:
@ivanyu I have addressed your most recent feedback, plus I've added a fix for the broken implementation of the Hash transformer

brbrown25 commented:
the broken build on integration seems to be caused by `com.github.dockerjava.api.exception.NotFoundException: {"message":"No such image: quay.io/testcontainers/ryuk:0.2.3"}`

brbrown25 commented:
@ivanyu I was able to fix the broken integration tests by bumping the test containers version.

ivanyu commented:
Thank you to the fix in `Hash`, @brbrown25! May I ask you to make a separate PR for this?

brbrown25 commented:
> May I ask you to make a separate PR for this?

Yes I'll do that this morning and then remove the commit from this pr!

brbrown25 commented:
Flattened and I moved the hash changes into a separate pr here https://github.com/aiven/aiven-kafka-connect-transforms/pull/31

brbrown25 commented:
@ivanyu I went ahead and removed the conditional logic from this one as well per your comment on the hash transformer.

ivanyu commented on README.md:
```diff
@@ -104,6 +104,33 @@ transforms.TombstoneHandler.type=io.aiven.kafka.connect.transforms.TombstoneHand
+ - `io.aiven.kafka.connect.transforms.ConcatField$Value` - works on values.
+
+The transformation defines the following configurations:
+ - `field.names` - A comma separated list of fields to concatenate.
```
```suggestion
 - `field.names` - A comma-separated list of fields to concatenate.
```

ivanyu commented on src/main/java/io/aiven/kafka/connect/transforms/ConcatFieldsConfig.java:
```diff
@@ -0,0 +1,87 @@
+final class ConcatFieldsConfig extends AbstractConfig {
+    public static final String FIELD_NAMES_CONFIG = "field.names";
+    private static final String FIELD_NAMES_DOC =
+        "A comma separated list of fields to concatenate.";
```
```suggestion
        "A comma-separated list of fields to concatenate.";
```

ivanyu reviewed (changes requested):
LGTM, just a couple of small things

ivanyu reviewed (approved):
LGTM, thanks @brbrown25
